rpcwebsocket.go

```diff
@@ -2035,802 +2079,146 @@ func deserializeOutpoints(serializedOuts []dcrjson.OutPoint) ([]*wire.OutPoint,
+// string slice.
+func rescanBlock(filter *wsClientFilter, block *dcrutil.Block) []string {
+	var transactions []string
 
-// unspentSlice returns a slice of currently-unspent outpoints for the rescan
-// lookup keys.  This is primarily intended to be used to register outpoints
-// for continuous notifications after a rescan has completed.
-func (r *rescanKeys) unspentSlice() []*wire.OutPoint {
-	ops := make([]*wire.OutPoint, 0, len(r.unspent))
-	for op := range r.unspent {
-		opCopy := op
-		ops = append(ops, &opCopy)
-	}
-	return ops
-}
-
-// ErrRescanReorg defines the error that is returned when an unrecoverable
-// reorganize is detected during a rescan.
-var ErrRescanReorg = dcrjson.RPCError{
-	Code:    dcrjson.ErrRPCDatabase,
-	Message: "Reorganize",
-}
-
-// Decred - TODO: This function needs to scan addresses/pks in both tx trees; right
-// now it only looks at the regular tx tree
-// rescanBlock rescans all transactions in a single block.  This is a helper
-// function for handleRescan.
-func rescanBlock(wsc *wsClient, lookups *rescanKeys, blk *dcrutil.Block,
-	parent *dcrutil.Block) {
-	txTreeRegularValid := dcrutil.IsFlagSet16(blk.MsgBlock().Header.VoteBits,
-		dcrutil.BlockValid)
-
-	// No need to rescan tx from genesis block.
-	if parent == nil {
-		return
-	}
-
-	var allTransactions []*dcrutil.Tx
-
-	if txTreeRegularValid {
-		allTransactions = append(allTransactions, parent.Transactions()...)
-	}
-	allTransactions = append(allTransactions, blk.STransactions()...)
-
-	for _, tx := range allTransactions {
-		// Hexadecimal representation of this tx.  Only created if
-		// needed, and reused for later notifications if already made.
-		var txHex string
-
-		// All inputs and outputs must be iterated through to correctly
-		// modify the unspent map, however, just a single notification
-		// for any matching transaction inputs or outputs should be
-		// created and sent.
-		spentNotified := false
-		recvNotified := false
-
-		// Get the stake tx type.
-		txType := stake.DetermineTxType(tx.MsgTx())
-
-		for i, txin := range tx.MsgTx().TxIn {
-			// Skip stakebase.
-			if txType == stake.TxTypeSSGen && i == 0 {
-				continue
+	// Need to iterate over both the stake and regular transactions in a
+	// block, but these are two different slices in the MsgTx.  To avoid
+	// another allocation to create a single slice to range over, the loop
+	// body logic is run from a closure.
+	//
+	// This makes unsynchronized calls to the filter and thus must only be
+	// called with the filter mutex held.
+	checkTransaction := func(tx *wire.MsgTx, tree int8) {
+		// Keep track of whether the transaction has already been added
+		// to the result.  It shouldn't be added twice.
+		added := false
+
+		inputs := tx.TxIn
+		if tree == dcrutil.TxTreeRegular {
+			// Skip previous output checks for coinbase inputs.  These do
+			// not reference a previous output.
+			if blockchain.IsCoinBaseTx(tx) {
```

Comment:
No, when it *is* a coinbase, then these checks can be skipped over.  Inputs are looped before outputs, and it jumps to outputs when it is a coinbase.